src/systems/camera-system.js

```diff
@@ -0,0 +1,182 @@
+import { waitForDOMContentLoaded } from "../utils/async-utils";
+import { setMatrixWorld } from "../utils/three-utils";
+import { paths } from "./userinput/paths";
+import { getBox } from "../utils/auto-box-collider";
+import qsTruthy from "../utils/qs_truthy";
+
+const calculateViewingDistance = (function() {
+  const center = new THREE.Vector3();
+  return function calculateViewingDistance(camera, object) {
+    const box = getBox(object.el, object.el.getObject3D("mesh"), true);
+    box.getCenter(center);
+    const halfYExtents = Math.max(Math.abs(box.max.y - center.y), Math.abs(center.y - box.min.y));
+    const halfXExtents = Math.max(Math.abs(box.max.x - center.x), Math.abs(center.x - box.min.x));
+    const halfVertFOV = THREE.Math.degToRad(camera.el.sceneEl.camera.fov / 2);
+    const halfHorFOV =
+      Math.atan(Math.tan(halfVertFOV) * camera.el.sceneEl.camera.aspect) * (object.el.sceneEl.is("vr-mode") ? 0.5 : 1);
+    const margin = 1.05;
+    const l1 = Math.abs((halfYExtents * margin) / Math.tan(halfVertFOV));
+    const l2 = Math.abs((halfXExtents * margin) / Math.tan(halfHorFOV));
+    const l3 = Math.abs(box.max.z - center.z) + Math.max(l1, l2);
+    const l = object.el.sceneEl.is("vr-mode") ? Math.max(0.25, l3) : l3;
+    return l;
+  };
+})();
+
+const moveRigSoCameraLooksAtObject = (function() {
+  const cqInv = new THREE.Quaternion();
+  const owq = new THREE.Quaternion();
+  const v1 = new THREE.Vector3();
+  const v2 = new THREE.Vector3();
+  const owp = new THREE.Vector3();
+
+  return function moveRigSoCameraLooksAtObject(rig, camera, object) {
+    object.getWorldQuaternion(owq);
+    object.getWorldPosition(owp);
+    cqInv.copy(camera.quaternion).inverse();
+    rig.quaternion.copy(owq).multiply(cqInv);
+
+    v1.set(0, 0, 1)
+      .multiplyScalar(calculateViewingDistance(camera, object))
+      .applyQuaternion(owq);
+    v2.copy(camera.position).applyQuaternion(rig.quaternion);
+    rig.position.subVectors(v1, v2).add(owp);
+
+    rig.matrixNeedsUpdate = true;
+  };
+})();
+
+let numCameraModes = 0;
+export const CAMERA_MODE_FIRST_PERSON = numCameraModes++;
+export const CAMERA_MODE_THIRD_PERSON_NEAR = numCameraModes++;
+export const CAMERA_MODE_THIRD_PERSON_FAR = numCameraModes++;
+export const CAMERA_MODE_INSPECT = numCameraModes++;
+
+const m2 = new THREE.Matrix4();
+const m3 = new THREE.Matrix4();
+const offset = new THREE.Vector3();
+export class CameraSystem {
+  constructor() {
+    this.mode = CAMERA_MODE_FIRST_PERSON;
+    waitForDOMContentLoaded().then(() => {
+      this.playerCamera = document.getElementById("player-camera");
+      this.playerRig = document.getElementById("player-rig");
+      this.cameraEl = document.getElementById("experimental-camera");
+      this.rigEl = document.getElementById("experimental-rig");
+      this.rigEl.object3D.matrixWorldNeedsUpdate = true;
```

Comment:
why is this here and not at the call site which modifies the matrix? also the idiomatic way to do this is just `.matrixNeedsUpdate = true`, the world flag is ideally just used internally (the `matrixNeedsUpdate` flag should deal with the world and local cases)

Reply:
after reading the rest of the code, I think you might misunderstand the contract of these various bits. here's the contract (might want to verify with @netpro2k, even though I wrote the code I tend to forget it ;))

- If you update the position/rotation/scale of an object3D, you *must* set `matrixNeedsUpdate = true`
- You shouldn't ever have to touch `matrixWorldNeedsUpdate`
- If you're using `getWorldPosition`, etc, you don't need to do anything special, just use those functions
- If you're reading `matrix` or `matrixWorld` directly, you need to call `updateMatrices()` on the object first

Reply:
Yep, corrected these